### PR TITLE
Add some documentation about descriptive task input and output types

### PR DESCRIPTION
**Pull request by jjohannes**
See: #2629

### CODE REVIEW

File: subprojects/docs/src/samples/userguide/tasks/descriptiveInputsAndOutputs/build.gradle [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    void setBinaries(FileCollection binaries) {
+        this.binaries.setFrom(binaries)
+    }
+
+    void setDist(Provider<RegularFile> dist) {
+        this.dist.set(dist)
+    }
+
+    @TaskAction
+    void doPackageWork() throws IOException {
```
Comment (bmuschko) [inline subprojects/docs/src/samples/userguide/tasks/descriptiveInputsAndOutputs/build.gradle]: :o: In Groovy you won't have to declare the `IOException`.

File: subprojects/docs/src/samples/userguide/tasks/descriptiveInputsAndOutputs/build.gradle [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    void setExecutable(Provider<RegularFile> executable) {
+        this.executable.set(executable)
+    }
+    void setInstallDirectory(Provider<Directory> installDirectory) {
+        this.installDirectory.set(installDirectory)
+    }
+
+    @TaskAction
+    void doInstallWork() throws IOException {
```
Comment (bmuschko) [inline subprojects/docs/src/samples/userguide/tasks/descriptiveInputsAndOutputs/build.gradle]: :o: In Groovy you won't have to declare the `IOException`.

File: subprojects/docs/src/docs/userguide/customPlugins.adoc [hunk 3]
```diff
@@ -105,30 +105,54 @@ The following demonstrates the usage of the type for mapping an extension proper
...
+execution time). As an example consider the following plugin with `compile`, `package` and `install` tasks which are wired via inputs and outputs:
+
+++++
+<sample id="descriptiveInputsAndOutputs" dir="userguide/tasks/descriptiveInputsAndOutputs" includeLocation="true" title="Descriptive input and output file declarations">
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+In the example above, the task outputs and inputs are connected before any concrete files system location is defined. This is possible because file locations
+are described by using providers that are created using `newOutputFile()` and `newOutputDirectory()` for outputs as well as `newInputFile()` and `files()` for inputs.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/customPlugins.adoc]: :x: We should add a link to the DSL guides for those methods.

File: subprojects/docs/src/docs/userguide/customPlugins.adoc [hunk 4]
```diff
@@ -105,30 +105,54 @@ The following demonstrates the usage of the type for mapping an extension proper
...
+Note how the corresponding setters work with providers instead of setting concrete file locations. This way the files are only actually resolved when they are
+needed during execution. The setters can be called at any time before the task is executed and the change will automatically affect all related input and output properties.
+
+When configuring the locations, `project.layout.buildDirectory` is used as entry point, which is another provider. This places all files relative to the project's build directory location -
+independent of if and when the build directory is changed during build configuration.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/customPlugins.adoc]: :x: "of the fact"?

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/customPlugins.adoc]: "build configuration" - Do we mean the configuration phase of the build lifecycle here?

**Comment by oehme** [comment]
@adammurdoch since you requested to put this on hold, can you please pick this back up once the API is good enough?

**Comment by bmuschko** [comment]
@adammurdoch Is this something that's ready to be documented for 4.3?

**Comment by jjohannes** [comment]
Superseded by #3478.

**Pull request closed without merging**